### ISSUE TITLE
[feature request] disable harmful broadcast

### ISSUE THREAD

**Issue by raingo** [tensorflow/tensorflow#1406]
Broadcast may lead to hidden bugs. For example, the following code,

```
xent = tf.nn. sigmoid_cross_entropy_with_logits(logits, y)
loss = tf.reduce_mean(xent)
```

When `logits` is of the shape `(200,1)` and the `y` is of the shape `(200,)`, the xent becomes `(200, 200)`, which is wrong. What's worse, the `reduce_mean` will hide this problem by reducing everything into a scalar.

Request,
1. more shape validation in `sigmoid_cross_entropy_with_logits`
2. option to totally disable broadcast

**Comment by vrv** [comment]
1. is done: https://github.com/tensorflow/tensorflow/blob/3871973b302b3c8357a4984fb37d934cf4633f21/tensorflow/python/ops/nn.py#L287 -- will be in our next binary release.

**Comment by raingo** [comment]
Thanks. I guess the second is impossible. so closing.

**Comment by vrv** [comment]
Nothing is impossible, it would just take a lot of work to plumb that through everywhere, to set the option globally, and for everything to work as well without broadcasting.  For example, when computing gradients, automatic broadcasting is quite useful to simplify gradient expressions, particularly when the shapes are unknown at construction time.  I've seen similar requests made for numpy, and it sounds like higher level libraries typically provide the 'safety' for those users who want it.  However, I'm open to the idea if it can be done in a nice way.

**Comment by rasmusbergpalm** [comment]
I have spent 8 hours on exactly this bug. An option to disable broadcasting globally or raising a warning, etc. would be very nice.
